### DISCUSSION TITLE
`refetchOnWindowFocus` not working in v5

### DISCUSSION THREAD

**Discussion by mamlzy** [TanStack/query#6568] [Q&A] [answered]
I'm using nextjs 14 app router, here is my react-query provider:

<img width="446" alt="image" src="https://github.com/TanStack/query/assets/63343105/dabd952b-be81-481c-b1ca-407ce66fcaa0">

and here is my custom hooks useQuery example:

<img width="453" alt="image" src="https://github.com/TanStack/query/assets/63343105/086c4f42-342f-49bc-9cf5-d5700cd25500">

i have another project using v4 and it's working fine. after switching browser tab, it shows refetch on the network tab. I don't now what's the problem here, i need some help, thanks!

**Comment by TkDodo** [comment, marked as answer]
have you seen the changes around window focus refetching?

https://tanstack.com/query/v5/docs/react/guides/migrating-to-v5#window-focus-refetching-no-longer-listens-to-the-focus-event

**Reply by mnpenner** [reply]
Hah! I use multiple monitors and so it always used to refetch when I switched from my IDE back to my browser and now in v5 it doesn't. I thought it was broken. This makes sense though, thanks for the link!

**Reply by gandhis1** [reply]
Reading the motivations in https://github.com/TanStack/query/pull/4805, it seems like it wasn't intentional that multiple monitor workflows were affected. I had previously relied on people clicking away -> clicking back to trigger a refresh. I ended up using the code excerpt in this PR to restore the original behavior, but I do wonder whether there's a better way to handle that by default.

**Comment by anselvo** [comment]
Hi, tbh it is really hard to find this change in migration docs, because when I was migrating from v4 to v5 I've notice changes and when I was trying to find it in the docs I searched for the flag and flag was noticed only once not in that place, so I though something was broken) It would be nice to see migration guide with focusListener in the migration guide. Probably I would add info about how to archive full window focus to this page as well https://tanstack.com/query/v5/docs/framework/react/guides/window-focus-refetching

P.S. it is really sad that project doesn't support `focus` any more, because this was exact behaviour that I've expected and switching tabs looks like a bug for me because when I switch between windows it doesn't refetch data :( So, probably it should be renamed then or it would be nice to add support for focus, thanks)

**Reply by TkDodo** [reply]
You can pass a custom event listener and listen to the focus event if you like. Please read the focusManager docs

**Reply by anselvo** [reply]
Yeah I know that what I did, I have couple of points;
1. name `onWindowFocus` is confusing because I expect that event `focus` will work in this case, moreover, I expect that word `window` is referring to browser window not a browser tab. Probably it should be called like `refetchOnTabFocus`
2. I would suggest support two listeners and add one more option, because right now I cannot easily separate these two listeners from one option `refetchOnWindowFocus` for example if I want to give client a chance to choose when he wants refetch request or I want some requests work on window focus and some of them on tab

Please, consider this improvements :bow:

**Reply by brauliodiez** [reply]
Just to avoid misunderstoods and be more accurate, should be `refetchOnWindowsFocus` marked as deprecated and create a new prop called `refetchOnVisibilityChange`?

**Reply by mnpenner** [reply]
> Just to avoid misunderstoods and be more accurate, should be `refetchOnWindowsFocus` marked as deprecated and create a new prop called `refetchOnVisibilityChange`?

`refetchOnVisibilityGain` or something. "change" implies both when it loses visibility and gains visibility; I'm assuming refetch only happens on the latter.

I'd like to see options for both though. The on focus behavior was quite nice for development (when I switch between monitors, there's no visibility change)

**Reply by namdao2000** [reply]
Hi, I'm stumbling upon this now, how can I achieve the old behaviour where it refetches when switching tabs?

**Reply by TkDodo** [reply]
with a custom event handler that listens to the `focus` event: https://tanstack.com/query/v5/docs/framework/react/guides/window-focus-refetching#custom-window-focus-event

**Reply by yoshegg** [reply]
In case some stumbles on this thread: As I also wanted to have refetch when refocusing the window on OS level (and not the tab), I added a listener to the `focus` event:

```javascript
focusManager.setEventListener((handleFocus) => {
    // Listen to visibilitychange or focus
    if (typeof window !== 'undefined' && window.addEventListener) {
      const visibilitychangeHandler = async () => {
        if (document.visibilityState === 'visible' || document.hasFocus()) {
          await queryClient.refetchQueries({ stale: true })
        }

        // handleFocus(document.visibilityState === 'visible' || document.hasFocus())
      }
      window.addEventListener('visibilitychange', visibilitychangeHandler, false)
      window.addEventListener('focus', visibilitychangeHandler)
      return () => {
        // Be sure to unsubscribe if a new handler is set
        window.removeEventListener('visibilitychange', visibilitychangeHandler)
        window.removeEventListener('focus', visibilitychangeHandler)
      }
    }
  })
```

I did not understand what the parameter / function `handleFocus` does, but it also works without; at least in a way that I need it to work.

**Reply by mnpenner** [reply]
Why'd you end up going with `queryClient.refetchQueries({ stale: true })` instead of `handleFocus(document.visibilityState === 'visible' || document.hasFocus())` ?

**Reply by yoshegg** [reply]
Using `handleFocus` does not work for me, because it only refetches once when the focus event happens. 

To reproduce: 
1. Open window and show DevTools network requests
2. Click outside of window, wait for `staleTime` to run out, click into window again
    - Network Requests are done again
3. Click outside of window again, wait for `staleTime` to run out, click into window again
    - No network requests are done

When using `queryClient.refetchQueries({ stale: true })` it works consistently every time.

**Reply by juanrgon** [reply]
I was able to get this to work by calling `handleFocus` on window focus and blur:

```tsx
focusManager.setEventListener((handleFocus) => {
  const setIsFocused = () => { handleFocus(true) };
  const setIsNotFocused = () => { handleFocus(false) };

  if (typeof window !== "undefined" && window.addEventListener) {
    window.addEventListener("focus", setIsFocused, false);
    window.addEventListener("blur", setIsNotFocused, false);

    return () => {
      // Be sure to unsubscribe if a new handler is set
      window.removeEventListener("focus", setIsFocused);
      window.removeEventListener("blur", setIsNotFocused);
    };
  }
});
```

**Comment by yairhaimo** [comment]
My problem was that my dev tools were open and the "emulate focused page" option was on so the page never went **out** of focus in order to return to be **in** focus and refetch.

I wanted to put it out there because maybe someone else will run into this issue.

**Reply by xanderknox** [reply]
honestly its embarrassing the amount of time i just spent trying to figure out why it just stopped working suddenly.... thank u

**Reply by TkDodo** [reply]
would you want to add this as a “gotcha” to the docs?

https://tanstack.com/query/latest/docs/framework/react/guides/window-focus-refetching

**Reply by brandongit2** [reply]
Oh. My. God. Never in a million years could I have figured that out myself...

**Reply by ptts** [reply]
First thought you were talking about the TanStack Query Devtools so I was trying to find that setting 😄

But you were talking about Chrome Devtools! 💡
https://developer.chrome.com/docs/devtools/rendering/apply-effects#emulate_a_focused_page

**Reply by cnorthfield** [reply]
What a hero. Would have never remembered I enabled that.
